### PR TITLE
[fix] fix GTP+DCP ckpt saving/loading for GDP module

### LINKED ISSUE #6501: [feat] GTP+DCP
GTP+DCP(Distributed Checkpoint) has been supported in the GTP init MR #4967 

This issue will keep tracking following-up bug-fixes / perf optimizations for GTP+DCP related stuffs.
- fix GTP+DCP for GDP moudle (#6503 )
- fix GTP+DCP when parallel_state is not initialized (#6775 )
- fix the case when some params is sharded across GTP while been duplicated across TP group (e.g., the latent gemm) (#6781 )
- GTP+DCP+fp8-param save->load is not bitwise matching 
        - MXFP8 (bytes + scale_A) → dequantize → BF16 → save → load → quantize → MXFP8 (bytes + scale_B)
               - Where scale_A (from fp32 master w) ≠ scale_B
               - quantize(dequantize(mfxp8) is a lossy round-trip, not identical
        - solution: during loading: cast from fp32 main params into mxfp8 params instead (related MR: #6666 )
- Fix GTP checkpoint load failing across different GTP degrees (#7293)

### PR DESCRIPTION
**Pull request by fanshiqing**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

fix(gtp): gather GDP in_proj shards before the DCP split
    
    GatedDeltaProductMixer.sharded_state_dict asserted under GTP_remat:
    in_proj.weight is GTP-sliced on axis 0 and pad-aligned, so the local
    shard matches neither in_proj_dim nor the [z|V|K|Q|b|a] split
    boundaries.
    
    Port MambaMixer's treatment -- all-gather the shards, strip the pad,
    then split (checkpoint stays identical to a non-GTP_remat run); wrap
    merge_fn to re-pad and re-slice on load.

### NT4-proxy save/load test at 100th iter

image

### UTs
Full GTP suite is green — overall_rc=0, 17/17 files, 122 passed, 0 failed 
image

## Issue tracking

Linked issue: Related to #6501 

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### CODE REVIEW

**Review by deepakn94 — APPROVED** [review]

File: tests/unit_tests/generalized_tensor_parallel/test_gtp_dcp.py [hunk 1]
```diff
@@ -942,6 +942,261 @@ def _worker_mamba_inproj_optim_param_map(rank, world_size, port):
...
+        # Pre-fix this raised AssertionError((in_proj_dim, ShardedTensor(...))).
+        sd = mixer.sharded_state_dict(prefix='mixer.', metadata=metadata)
+
+        factory = sd['mixer.in_proj.weight']
+        assert isinstance(factory, ShardedTensorFactory), type(factory)
+        # Save side: the gathered tensor is the full TP-local width, pad stripped.
+        assert factory.data.size(0) == in_proj_dim, (factory.data.size(0), in_proj_dim)
+
+        chunks = factory.build_fn(factory.key, factory.data, factory.replica_id, None)
+        assert [t.key.rsplit('.', 1)[-1] for t in chunks] == ["z", "V", "K", "Q", "b", "a"]
```
[the remark this answers is not in the document]
Comment (fanshiqing) [reply tests/unit_tests/generalized_tensor_parallel/test_gtp_dcp.py]: Done

**Review by Phlip79 — APPROVED** [review]

**Comment by Phlip79** [comment]
Thank you for adding tests!

**Review by Phlip79 — APPROVED** [review]
